services/s3/src/test/java/software/amazon/awssdk/services/s3/handlers/PutObjectHeaderTest.java

```diff
@@ -0,0 +1,100 @@
+import static com.github.tomakehurst.wiremock.client.WireMock.postRequestedFor;
+import static com.github.tomakehurst.wiremock.client.WireMock.putRequestedFor;
+import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlMatching;
+import static com.github.tomakehurst.wiremock.client.WireMock.verify;
+import static software.amazon.awssdk.http.Headers.CONTENT_TYPE;
+
+import com.github.tomakehurst.wiremock.junit.WireMockRule;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import software.amazon.awssdk.core.auth.AwsCredentials;
+import software.amazon.awssdk.core.auth.StaticCredentialsProvider;
+import software.amazon.awssdk.core.regions.Region;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.core.util.Mimetypes;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.testutils.RandomTempFile;
+
+public class PutObjectHeaderTest {
+
+    @Rule
+    public WireMockRule mockServer = new WireMockRule(0);
+
+    private S3Client s3Client;
+
+    private PutObjectRequest putObjectRequest;
+
+    @Before
+    public void setup() {
+        s3Client = S3Client.builder()
+                .credentialsProvider(StaticCredentialsProvider.create(AwsCredentials.create("akid", "skid")))
+                .region(Region.US_WEST_2).endpointOverride(URI.create(getEndpoint()))
+                .build();
+        putObjectRequest = PutObjectRequest.builder().bucket("test").key("test").build();
+    }
+
+    private String getEndpoint() {
+        return "http://localhost:" + mockServer.port();
+    }
+
+    @Test
+    public void putObjectBytes_headerShouldContainContentType() {
+        stubFor(any(urlMatching(".*"))
+                    .willReturn(aResponse()
+                                    .withStatus(200)
+                                    .withBody("{}")));
+        s3Client.putObject(PutObjectRequest.builder().bucket("test").key("test").build(), RequestBody.of("Hello World".getBytes()));
+        verify(putRequestedFor(anyUrl()).withHeader(CONTENT_TYPE, equalTo(Mimetypes.MIMETYPE_TEXT_PLAIN)));
+    }
+
+    @Test
+    public void putObjectFile_headerShouldContainContentType() throws IOException {
+        File file = new RandomTempFile("test.html", 10);
+        stubFor(any(urlMatching(".*"))
+                    .willReturn(aResponse()
+                                    .withStatus(200)
+                                    .withBody("{}")));
+        s3Client.putObject(putObjectRequest, RequestBody.of(file));
+        verify(putRequestedFor(anyUrl()).withHeader(CONTENT_TYPE, equalTo("text/html")));
+        file.delete();
+    }
+
+    @Test
+    public void putObjectUnknownFileExtension_contentTypeDefaultToBeStream() throws IOException {
```

Comment:
Can we add some tests to assert that the ContentType on the actual PutObjectRequest is honored over derived ones and also that content type added to raw headers in request config are honored.

Reply:
Good call! added